database/download.go

```diff
@@ -51,52 +68,42 @@ func (db *DB) DownloadCreate(ctx context.Context, user User, skylink Skylink) (*
 	return &up, nil
 }
 
-// DownloadsBySkylink fetches all downloads of this skylink
-func (db *DB) DownloadsBySkylink(ctx context.Context, skylink Skylink, offset, limit int) ([]Download, error) {
+// DownloadsBySkylink fetches a page of downloads of this skylink and the total
+// number of such downloads.
+func (db *DB) DownloadsBySkylink(ctx context.Context, skylink Skylink, offset, limit int) ([]DownloadResponseDTO, int, error) {
 	if skylink.ID.IsZero() {
-		return nil, errors.New("invalid skylink")
-	}
-	filter := bson.D{{"skylink_id", skylink.ID}}
-	opts := options.FindOptions{}
-	if offset > 0 {
-		opts.SetSkip(int64(offset))
-	}
-	if limit > 0 {
-		opts.SetLimit(int64(limit))
-	}
-	c, err := db.staticDownloads.Find(ctx, filter, &opts)
-	if err != nil {
-		return nil, err
+		return nil, 0, errors.New("invalid skylink")
 	}
-	downloads := make([]Download, 0)
-	err = c.All(ctx, &downloads)
-	if err != nil {
-		return nil, err
-	}
-	return downloads, nil
+	matchStage := bson.D{{"$match", bson.D{{"skylink_id", skylink.ID}}}}
+	return db.downloadsBy(ctx, matchStage, offset, limit)
 }
 
-// DownloadsByUser fetches all downloads by this user
-func (db *DB) DownloadsByUser(ctx context.Context, user User, offset, limit int) ([]Download, error) {
+// DownloadsByUser fetches a page of downloads by this user and the total number
+// of such downloads.
+func (db *DB) DownloadsByUser(ctx context.Context, user User, offset, limit int) ([]DownloadResponseDTO, int, error) {
 	if user.ID.IsZero() {
-		return nil, errors.New("invalid user")
+		return nil, 0, errors.New("invalid user")
 	}
-	filter := bson.D{{"user_id", user.ID}}
-	opts := options.FindOptions{}
-	if offset > 0 {
-		opts.SetSkip(int64(offset))
-	}
-	if limit > 0 {
-		opts.SetLimit(int64(limit))
+	matchStage := bson.D{{"$match", bson.D{{"user_id", user.ID}}}}
+	return db.downloadsBy(ctx, matchStage, offset, limit)
+}
+
+// downloadsBy fetches a page of downloads, filtered by an arbitrary match
+// criteria. It also reports the total number of records in the list.
+func (db *DB) downloadsBy(ctx context.Context, matchStage bson.D, offset, limit int) ([]DownloadResponseDTO, int, error) {
+	cnt, err := count(ctx, db.staticDownloads, matchStage)
+	if err != nil || cnt == 0 {
+		return []DownloadResponseDTO{}, 0, err
 	}
-	c, err := db.staticDownloads.Find(ctx, filter, &opts)
+	pipeline := generateUploadsDownloadsPipeline(matchStage, offset, limit)
+	c, err := db.staticDownloads.Aggregate(ctx, pipeline)
 	if err != nil {
-		return nil, err
+		return nil, 0, err
 	}
-	downloads := make([]Download, 0)
+	downloads := make([]DownloadResponseDTO, 0)
```

Comment:
I guess you could initialize this with the count now. Not sure what happens if between the count and the query a document got inserted or removed 🤔

Reply:
Yes, nice catch. A tiny optimisation but a free one. Except that I shouldn't init it with count (which is the total count of objects in the DB) but with limit.

If the number of records in the DB changes between the two DB requests we're going to have a bit of a dicrepancy, I guess, i.e. `count: 5, items: <arr with len 6>` or something like that. It doesn't seem like a critical thing in this particular case. We could counter that with a transaction or something like that but I'd rather keep it simple for the moment.